### PR TITLE
(feat): session utility

### PR DESCRIPTION
**Pull request by goosewobbler**
Building out the standalone feature proposed in #412.

TODO:
- [x] Fix mac error
- [x] Ensure it works with CJS
- [x] Check orphan window behaviour without `app.quit` on Linux & Windows
- [x] Fix windows EB error on CI
- [x] Units
- [x] Clean up
- [x] Docs
- [x] Fix windows E2E timeouts

### CODE REVIEW

**Comment by goosewobbler** [comment]
Seems to work on windows and linux, Mac builds broken with the following: 
```
 2024-04-09T12:46:44.674Z ERROR webdriver: unknown error: unknown error: Failed to create Chrome process.
    at getErrorFromResponseBody (file:///Users/runner/work/wdio-electron-service/wdio-electron-service/example/node_modules/webdriver/build/utils.js:195:12)
    at NodeJSRequest._request (file:///Users/runner/work/wdio-electron-service/wdio-electron-service/example/node_modules/webdriver/build/request/index.js:193:23)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async startWebDriverSession (file:///Users/runner/work/wdio-electron-service/wdio-electron-service/example/node_modules/webdriver/build/utils.js:64:20)
    at async Function.newSession (file:///Users/runner/work/wdio-electron-service/wdio-electron-service/example/node_modules/webdriver/build/index.js:20:45)
    at async remote (file:///Users/runner/work/wdio-electron-service/wdio-electron-service/example/node_modules/webdriverio/build/index.js:45:22)
    at async init (file:///Users/runner/work/wdio-electron-service/wdio-electron-service/example/node_modules/wdio-electron-service/src/session.ts:26:19)
    at async file:///Users/runner/work/wdio-electron-service/wdio-electron-service/example/e2e-standalone/api.spec.ts:24:17
```

**Comment by goosewobbler** [comment]
@christian-bromann before I add docs & units for this, does it cover the feature request?  The utility is called `startSession`, do we want to standardise with WDIO and call it `remote`?

File: src/cjs/session.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    [CUSTOM_CAPABILITY_NAME]: opts,
+  };
+
+  const launcher = new CJSElectronLauncher(opts, capabilities, testRunnerOpts);
+  const service = new CJSElectronService(opts);
+
+  await launcher.onPrepare(testRunnerOpts, [capabilities]);
+
+  console.log('setting caps', capabilities);
+
```
Comment (christian-bromann) [inline src/cjs/session.ts:21]:
```suggestion
```

File: src/session.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    [CUSTOM_CAPABILITY_NAME]: opts,
+  };
+
+  const launcher = new ElectronLaunchService(opts, capabilities, testRunnerOpts);
+  const service = new ElectronWorkerService(opts);
+
+  await launcher.onPrepare(testRunnerOpts, [capabilities]);
+
+  console.log('setting caps', capabilities);
+
```
Comment (christian-bromann) [inline src/session.ts:22]:
```suggestion
```

File: src/cjs/session.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { remote } from 'webdriverio';
+
+import { CJSElectronLauncher, CJSElectronService } from './classes.js';
+import { CUSTOM_CAPABILITY_NAME } from './constants.js';
+import { ElectronServiceOptions } from './types.js';
+import { Options } from '@wdio/types';
+
+export async function init(opts: ElectronServiceOptions) {
```
Comment (christian-bromann) [inline src/cjs/session.ts:8]: How about dynamically importing the ESM module here to avoid duplication?

**Review by christian-bromann — COMMENTED** [review]
I have no strong opinions how the function should be called. Honestly `remote` is also just an historical artifact which we could find a better name for. I would say, to make it very explicit and remove any doubts on what the user gets, I would call the function `startElectron` instead of `startSession`.

File: src/cjs/session.ts (hunk 3 above)
Comment (goosewobbler) [reply src/cjs/session.ts:8]: It causes problems with rollup on the `example` E2Es: https://github.com/webdriverio-community/wdio-electron-service/actions/runs/8908865877/job/24465227351

I'll see if I can remove the bundling - we're not bundling the `example-cjs` app but the other two proved difficult.

Comment (goosewobbler) [reply src/cjs/session.ts:8]: I removed rollup for the `example` (ESM) main, trying different things for preload - which needs to be CJS.  Now getting "exports is not defined in ES module scope" for the service as imported by `@wdio/utils` safeImport.  This happens regardless of bundling or lack thereof in `example`, so it must be a result of the dynamic import here.

Comment (goosewobbler) [reply src/cjs/session.ts:8]: I kept the bundling for preload scripts in the ESM examples - converting them to CJS syntax didn't help.  I can't see a way of removing the duplication here with `session.ts`.

**Comment by goosewobbler** [comment]
Also need to work out where this EB error is coming from on windows, occuring on CI since the merge with `main`: https://github.com/webdriverio-community/wdio-electron-service/actions/runs/8910525093/job/24469967733?pr=521

File: src/cjs/session.ts (hunk 3 above)
Comment (christian-bromann) [reply src/cjs/session.ts:7]: ok, we can keep it this way but let's add a comment documenting why the code is duplicated.

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -32,16 +32,16 @@ jobs:
...
-      - uses: actions/setup-node@v3
+      - uses: actions/setup-node@v4
         with:
-          node-version: 16.x
-      - uses: pnpm/action-setup@v2
+          node-version: 18.x
+      - uses: pnpm/action-setup@v3
         with:
-          version: 8.10.0
+          version: 9.0.6
```
Comment (christian-bromann) [inline .github/workflows/release.yml:44]: We could have dependabot do this job for us 😉

File: example-cjs/e2e-standalone/api.spec.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+  const appVersion = await browser.electron.execute((electron) => electron.app.getVersion());
+  if (appVersion !== packageJson.version) {
+    throw new Error(`appVersion test failed: ${appVersion} !== ${packageJson.version}`);
+  }
+
+  // Clean up - quit the app as it remains loaded
+  // Linux uses xvfb on CI, app.quit will break the E2Es in this case
+  if (!isCI) {
+    await browser.electron.execute((electron) => electron.app.quit());
```
Comment (christian-bromann) [inline example-cjs/e2e-standalone/api.spec.ts:48]: Would `browser.deleteSession()` do the same here?

**Review by christian-bromann — COMMENTED** [review]
Overall looks good to me. Great work! One question on how to end a session as well as an opportunity to automate pipeline updates.

File: .github/workflows/release.yml (hunk 4 above)
Comment (goosewobbler) [reply .github/workflows/release.yml:44]: Yeah, I'll add a ticket to look at that

**Review by christian-bromann — APPROVED** [review]
LGTM 🎉

File: example-cjs/e2e-standalone/api.spec.ts (hunk 5 above)
Comment (goosewobbler) [reply example-cjs/e2e-standalone/api.spec.ts]: Yes and it works with xvfb on CI 😄

**Pull request merged**